### ISSUE TITLE
[V10] auth.params.state has to be set when creating a new lock

### ISSUE THREAD

**Issue by saikat** [auth0/lock#488]
Right now, to set state on an Auth0Lock, you have to do:

````
this.lock = new Auth0Lock(this.clientId, this.domain, {
      auth: {
        redirect: true,
        redirectUrl: 'http://localhost:3000/',
        responseType: 'token',
        params: {
          state: 'next_url'
        }
      }
    })
  // Add callback for lock `authenticated` event
  this.lock.on('authenticated', this._doAuthentication.bind(this))
````

However, in the common case of wanting to redirect to the original URL after successful auth, you won't actually know the state param until calling lock.show().  Would it not make more sense to provide the state as an option to lock.show than when creating the lock?

**Comment by gnandretta** [comment]
Hi, we would like to keep the options in just one place. Can your use case be solved by storing the url in `localStorage` when you show lock, and then retrieve it when the authentication is done?

**Comment by KCraw** [comment]
How would you recommend using this with `auth: { responseType: 'code', redirect: true }`?

Currently we are using a SPA that handles server-side pre-rendering and Lock v9. To keep things consistent, we are using redirect mode with a response type of code that redirects to an auth endpoint. Our server uses the code to request the JWT and uses a redirect to set the JWT in an HTTPOnly cookie. We were using `state` to allow redirecting to the original URL where the user triggered Lock.

**Comment by KCraw** [comment]
Not sure when Lock actually uses the settings, but is it possible to allow passing a thunk for `options.auth.params`?  This way it can there's still a single options object passed during Auth0 creation, but params can still be lazily evaluated just before things transition to set any params that might be needed server side, like the user's current location. 

Otherwise, I can't think of a way to use server-side pre-rendering of a SPA with Lock.

**Comment by saikat** [comment]
Kristopher, I actually have a server-side pre-rendered SPA with Lock.  I'm
just using the Lock as UI and not using the lock's authenticated handler to
handle the callback.  Basically I'm using a serverside route to handle the
callback, which uses auth0-passport along with cookie-session to set the
token as a session cookie.  With this method, I just create the lock and
call show on it every time you want to show a login.

**Comment by KCraw** [comment]
@saikit That's similar to my own setup, though we're not using passport. But without the ability to pass arbitrary data to the backend, how are you handling SSR after login?

Previously with v9 we were able to pass the current location to the server, so the server could redirect and render the user's previous location.  

With the solution @gnandretta suggested, we'd have to render a generic page and wait for client-side JS to read from local storage and trigger a client-side redirect. That's how I'm handing it with v10, but I don't like it. 

I was requesting the ability to use thunks in the options object if the author's of Lock want to have a single options object passed into the constructor. But maybe that's not necessary.

Did you find a workaround to the problem in your original post?

**Comment by saikat** [comment]
You can use the state param to pass arbitrary data in the callback, and I'm
setting the JWT in a session cookie so I have access to it on first load of
the page.

**Comment by gnandretta** [comment]
@saikat @KCraw I suggested using `localStorage` because I didn't know you were rendering on the server.  The strategy of @saikat works best for this scenario but it seems a bit cumbersome so it would be better if we just allow to override some options when showing lock.

**Comment by saikat** [comment]
I'm not sure if there is a way to make SSR work correctly without putting
the JWT in a session cookie.  The fundamental issue is that you need access
to the JWT in order to know what to render when the first request hits your
web server.  So the browser needs to send information on your session with
that request, which I believe is only possible if the session info is in a
cookie.

The lock could be configured to set that cookie, though, instead of storing
the JWT in local storage. This is actually not even a part of the lock's
implementation right now -- just in the example, instead of doing
localStorage.set, you'd do cookie.set, and now your server would have
access to the session data for SSR.  The cookie wouldn't be able to be
HttpOnly then, but I don't think that makes the cookie any less secure to
use than localStorage already is (as in, both are vulnerable to XSS
attacks).

**Comment by KCraw** [comment]
@gnandretta  Thanks! Really I think it's probably just `state` that we need to be able to override when calling `lock.show()`.

@saikat  To be clear: state is passed as a query param to the callback URL when you use `responseType: 'code'`, so is available to the server. We're not using the Lock event handlers (or callback in v9). We handle the callback server-side and to trigger a server-to-server request for the JWT, and then set a HttpOnly cookie with the JWT. It sounds like you are doing pretty much the same thing, just using Passport. 

Previously we could pass state when calling `lock.show`, which allowed us to use `window.location` to pass the current location in the SPA, so that it would be available to the server via query params. After setting the session cookie, we trigger a redirect based on the state param rather than rendering a page (of course, the redirect results in a proper SSR with access to the JWT credentials). 

The only problem here is that we can no longer set the state that will be passed to our server when we show Lock. We could set it when initializing Auth0, but that tends to be a long lived object that survives SPA navigation. 

Again, I think we have the exact same problem and very similar setups. Though I'm not sure from your responses whether you've got a current setup with v10 that is allowing you to have `auth.params.state` always be in sync with where the user is in your SPA.

**Comment by chriswhong** [comment]
+1, I have an SPA and instantiate lock once, then pass in state to .show() so I can redirect the user to whatever page they were trying to view.  This doesn't work anymore with v10

**Comment by gnandretta** [comment]
Implemented in #550 and released in `v10.2.0`.
